### PR TITLE
MO436 sarsa lambda algorithm

### PR DESCRIPTION
**Pull request by sueny**
Implementations: 
 - Sarsa lambda agent
 - train script 

Summary:
This initial implementation is still random, the agent does not learn anything yet. There are some decisions we need to make regarding states before we move on.

You guys can review and then merge this PR. Or wait for the final working version. I created this PR to make the code available to al, thus you can build your code on top of it if needed.

PS: Hubert's profile wasn't suggested for me as a reviewer. I couldn't find his user. Hubert if you see this please add yourself as a reviewer 😃

### CODE REVIEW

File: project_RL/sarsa/train.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+            action = next_action
+            env.render()
+            print("a:", action, "i:", step, "reward:", reward, "info:", info)
+            if done:
+                print("done?", done, "total reward:", total_reward, "info:", info)
+            step += 1
+
+    env.close()
+
+
```
Comment (Ronnypetson) [inline project_RL/sarsa/train.py:50]: I suggest putting this part from line 51 on inside a "if __name__ == '__main__':"

File: project_RL/sarsa/sarsa_lambda_agent.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
+import random
+
+import numpy as np
+
+
+class SarsaLambdaAgent:
```
[the remark this answers is not in the document]
Comment (Ronnypetson) [reply project_RL/sarsa/sarsa_lambda_agent.py:6]: Sounds like a good idea. I think first everybody needs to have clear in mind what is an agent in this context, so that we can abstract and create the super class.

File: project_RL/sarsa/sarsa_lambda_agent.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+        if random.random() < self.epsilon:
+            return random.choice(range(self.action_size))
+        else:
+            return np.argmax(q_state)
+
+    def train(self, state, action, reward, new_state, new_action, done):
+        """Updates the action value for every pair state, action
+        in proportion to TD-error and eligibility trace
+
+        """
```
Comment (Ronnypetson) [inline project_RL/sarsa/sarsa_lambda_agent.py:48]: From what I saw in the Sutton's book (page 185), this seems to going in the right way.

**Review by caiolimaa — APPROVED** [review]

File: project_RL/sarsa/sarsa_lambda_agent.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+class SarsaLambdaAgent:
+    """Sarsa Lambda Agent exploits and explores an environment
+    in order to converge to the optimal action-value function
+    """
+
+    def __init__(self, env, discount_rate=0.9, learning_rate=0.1, epsilon=0.5):
+        self.action_size = env.action_space.n
+
+        # TODO: decide together in group what we will store as state
```
[the remark this answers is not in the document]
Comment (sueny) [reply project_RL/sarsa/sarsa_lambda_agent.py:14]: Our state will be composed of the following variables:
dlw - distance to left wall
drw - distance to right wall
duw - distance to upper wall
dbw - distance to bottom wall
dir - direction
dxg - distance to goal in the x axis
dyg -  distance to goal in the y axis

# dlw, drw, duw, dbw, dxg, dyg
# 0, -1, -1, 5, 3

in which, 0 means the agent is side by side with the object
-1 unkwown, i.e. the object is not visible in the agent partial view  

Therefore, the state size is: dlw x drw x duw x dbw x dxg x dyg
# 7 x 7 x 7 x 7 x 4 x 7 x 7  = 823543 

q value table is equal to state size x number of actions = 823543 x 7 = 5764801

**Review by Ronnypetson — APPROVED** [review]
